### ISSUE TITLE
SOQL query using autofetch=true drops a batch of results set when using queryMore after the initial query

### ISSUE THREAD

**Issue by peternhale** [jsforce/jsforce#1385]
I ran into an issue while using query, with autoFetch true and then wanting to fetch more records using queryMore.

As I understand autoFetch, once the query call finishes the response will contain the properties done, totalSize, records[] and optionally nextRecordsUrl. The SOQL query I am using returns more records than maxFetch, with the properties:
* done = false
* totalSize = number of records in my query
* records[maxFetch is size]
* nextRecordsUrl= URL to the next batch of 2000 records.

At this moment I use queryMore function to start getting the remaining records. What I am finding is that with autoFetch = true, is one batch will be missing and I believe it is the batch that follows the number of records specified by maxFetch.

For example:
* SOQL query returns 20,000 records
* maxFetch set to 10,000 records
* after following all presented nextRecordsUrls the total number of records is 18,000

I am using the most recent version of v2@beta and node 18.18.*

I wrote a reproducer (see [gist](https://gist.github.com/peternhale/37b6681535888ade20673c55ddc94537)) that displays the key properties from the query. Both use maxFetch = 2,000. The first is with autoFetch false and as you can see all the queried rows are returned. Each query/queryMore result displays the number of the query/queryMore, batch size and `nextRecordsUrl`. Notice the first `nextRecordsUrl` ends with `-2000`.

```
Executing query: SELECT Id, Name FROM Account with autoFetch: false
Query count: 1 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-2000
Query count: 2 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-4000
Query count: 3 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-6000
Query count: 4 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-8000
Query count: 5 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-10000
Query count: 6 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-12000
Query count: 7 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-14000
Query count: 8 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-16000
Query count: 9 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-18000
Query count: 10 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivsRiHU6ACN-20000
Query count: 11 Records in this batch: 1 nextRecordsUrl: 
Total records: 20001
```
This next run is the same except with autoFetch = true. In this result you can see there is one fewer queries and the nextRecordsUrl for `-2000` is missing.
```
Executing query: SELECT Id, Name FROM Account with autoFetch: true
Query count: 1 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivvJCwGoACL-4000
Query count: 2 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivvJCwGoACL-6000
Query count: 3 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivvJCwGoACL-8000
Query count: 4 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivvJCwGoACL-10000
Query count: 5 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivvJCwGoACL-12000
Query count: 6 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivvJCwGoACL-14000
Query count: 7 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivvJCwGoACL-16000
Query count: 8 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivvJCwGoACL-18000
Query count: 9 Records in this batch: 2000 nextRecordsUrl: https://<redacted>/services/data/v50.0/query/0r8xx2PivvJCwGoACL-20000
Query count: 10 Records in this batch: 1 nextRecordsUrl: 
Total records: 18001
```

**Comment by nellicus** [comment]
I am seeing similar behaviour using `"jsforce": "^1.11.1",`

**Comment by cristiand391** [comment]
Thanks for the repro, Pete!

I can repro locally, when calling `conn.query` with `autoFetch: true` it gets the first 2000 records, then follows `nextRecordsUrl` and returns that batch (so, it misses the first 2000 records).

**Comment by cristiand391** [comment]
after going over this a bit more I think you can simplify your code and skip this bug:

say you have 5000 account records and run the next line:
```ts
await conn.query("SELECT Id, Name FROM Account") // returns 2000 records
```
that will only return 2000 records (API limit) and you could paginate over it with `queryMore` like in your gist **or** use autoFetch to let jsforce paginate over the results and return the whole array of records:

```ts
await conn.query("SELECT Id, Name FROM Account", { autoFetch: true }) // returns 5000 records
```

In your gist you set both `autoFetch: true` **and** `maxFetch: 2000` (then manually paginate with `queryMore`), which triggers that bug swallowing the first batch because of `maxFetch`.

TL;DR
use `autoFetch: true` to let jsforce paginate over results, set `maxFetch` to a higher limit if you expect this query to return a bunch of results (default value is `10_000`):
https://jsforce.github.io/document/#:~:text=NOTE%3A%20When%20maxFetch%20option%20is%20not%20set%2C%20the%20default%20value%20(10%2C000)%20is%20applied.%20If%20you%20really%20want%20to%20fetch%20more%20records%20than%20the%20default%20value%2C%20you%20should%20explicitly%20set%20the%20maxFetch%20value%20in%20your%20query.

**Comment by peternhale** [comment]
@cristiand391 That is the pattern we used after discovering this issue; auto fetch = false and query more for fetching remaining rows.

**Comment by cristiand391** [comment]
This should be fixed in v3, see: https://github.com/jsforce/jsforce/discussions/1529
